rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -649,14 +649,31 @@ NodeParameters::undeclare_parameter(const std::string & name)
     throw rclcpp::exceptions::ParameterImmutableException(
             "cannot undeclare parameter '" + name + "' because it is read-only");
   }
-  if (!parameter_info->second.descriptor.dynamic_typing) {
+  if (!parameter_info->second.descriptor.undeclarable) {
```

Comment:
This would break existing nodes that rely on this behavior. 
```suggestion
  if (!parameter_info->second.descriptor.undeclarable && !parameter_info->second.descriptor.dynamic_typing) {
```

Reply:
i second this, we want to maintain the consistent behavior for the user space. or if we change that, maybe warning notice would be useful for the application.